services/FeedbackService.ts

```diff
@@ -0,0 +1,66 @@
+import { Service } from 'typedi';
+import { EntityManager } from 'typeorm';
+import { InjectManager } from 'typeorm-typedi-extensions';
+import * as moment from 'moment';
+import { FeedbackModel } from '../models/FeedbackModel';
+import { UserModel } from '../models/UserModel';
+import Repositories, { TransactionsManager } from '../repositories';
+import { PublicFeedback, Feedback, Uuid, ActivityType } from '../types';
+import { UserError } from '../utils/Errors';
+import { Config } from '../config';
+
+@Service()
+export default class FeedbackService {
+  private transactions: TransactionsManager;
+
+  constructor(@InjectManager() entityManager: EntityManager) {
+    this.transactions = new TransactionsManager(entityManager);
+  }
+
+  public async getFeedback(canSeeAllFeedback = false, user: UserModel): Promise<PublicFeedback[]> {
+    const feedback = await this.transactions.readOnly(async (txn) => {
+      const feedbackRepository = Repositories.feedback(txn);
+      if (canSeeAllFeedback) return feedbackRepository.getFeedback();
+      return feedbackRepository.getFeedbackByUser(user);
+    });
+    return feedback.map((fb) => fb.getPublicFeedback());
+  }
+
+  public async submitFeedback(user: UserModel, feedback: Feedback): Promise<PublicFeedback> {
+    const feedbackObject = FeedbackModel.create({ ...feedback, user });
+    const addedFeedback = await this.transactions.readWrite(async (txn) => {
+      const feedbackRepository = Repositories.feedback(txn);
+
+      const latestFeedback = await feedbackRepository.getLatestFeedback(user);
+      if (latestFeedback) {
+        const oneWeekAfterFeedback = moment(latestFeedback.timestamp).add(1, 'week');
+        if (oneWeekAfterFeedback.valueOf() > moment.now()) {
+          const daysLeft = oneWeekAfterFeedback.diff(moment(), 'days', false) + 1;
+          throw new UserError(`You must wait ${daysLeft} days to submit more feedback`);
```

Comment:
I agree, but how should we go about granting points only once a week? Feedback submitted within a week of another should still be acknowledgeable but not be rewarded points, and we don't want to refresh the week duration on feedback that hasn't been rewarded. So we need a way of keeping track of the last feedback submitted that has been rewarded points so that we can reward/not reward accordingly. Maybe a `hasBeenRewarded` field on each feedback? So we can query the latest feedback that has been rewarded and check the week long duration from that?

Reply:
Boolean check for acknowledged feedback within a week of the feedback being acknowledged?

But if we're acknowledging it maybe we should just grant points every time. This is where "acknowledged" and "acknowledged and useful" distinctions might come in handy, but again, maybe just grant on all acknowledged feedback? Maybe an internal "acknowledged but not useful" that's publicly displayed as just "submitted"?

Reply:
I'm thinking of adding an enum `status(SUBMITTED, MARKED_AS_SPAM, ACKNOWLEDGED, REWARDED)`. When admin acknowledges a specific feedback, if it's been more than a week since the last REWARDED feedback, then it gets marked as REWARDED, otherwise its marked as ACKNOWLEDGED. I'll include MARKED_AS_SPAM too since we aren't restricting submission rates so spam by trolls is somewhat inevitable

Reply:
Honestly, I think we should just grant points on all acknowledgement. People submitting a lot of useful feedback should be rewarded not penalized. I think `status(SUBMITTED, MARKED_AS_SPAM/DISCARDED/IGNORED, ACKNOWLEDGED)` is fine, where any acknowledgement carries a point bonus and `IGNORED` is shown to the submitter as `SUBMITTED`.

Reply:
sounds good